### DISCUSSION TITLE
program Micropython board without using the USB connector

### DISCUSSION THREAD

**Discussion by Dandy-Don** [micropython/micropython#14484] [RP2040 / Pico]
Hello, all!

I could use some help, please.

I'm wondering if it is possible to program an RP2040 board without using the USB connector?

I have a few boards in mind:  Pi Pico, Adafruit ItsyBitsy RP2040, & the Adafruit Feather RP2040.  The board will be in a housing that will not allow access to the USB connector or the boot select button.....

I think I can use the SWD pins on the board to program them, but I think that only applies to programming in C.  Correct?

Is there a way to program the board in Micropython without using a USB connector?

Thanks in advance for any help!
--Don

**Comment by robert-hh** [comment]
You have to use the SWD pins to upload the firmware. As part of that, it can have embedded Python code, so called frozen bytecode. That will be bundled with the firmware when you build the firmware. Only main.py and boot.py cannot be embedded as frozen bytecode. But that could be changed in main.c.  If UART pins are available in our set-up you can have a REPL prompt at an UART. Then you can upload Python code during runtime, e.g. using mpremote.

**Reply by Dandy-Don** [reply]
Thanks, robert-hh!

I'm not sure how to embed Python code into the main.py code.  Can you explain that better, please?  Or provide a link to some sample code....?

I will be using UART pins to talk to the master board, a Raspberry Pi 4.  The Pi will control some GPIO on the Pico or Adafruit board, as well as display some ADC from that board.....

**Reply by robert-hh** [reply]
You just have to place you python code into the rp2/modules directory and then build the firmware again. Then it will be embedded into the firmware with the advantage, that  is does not use RAM at runtime and the startup time is much faster. Alternatively you can use the manifest.py mechanism to include your python modules into the firmware. Seehttps://docs.micropython.org/en/latest/reference/manifest.html.
To run your main code (not called main.py) at start-up without requiring a specific main.py, you could add in rp2/main.c a line like:
`pyexec_frozen_module("my_main.py", true);`
after e.g. line 196, between running main.py and before the REPL loop.

**Reply by Dandy-Don** [reply]
Not sure if this is what I need to do to program the board without using USB.... Looks like it involves using C, not Micropython.

I'd really like to use the SWD pins for programming, but I'm not sure how....

Any other solutions?

**Comment by victorallume** [comment]
Are you using a wifi network with your boards? If you can get the webrepl running in your firmware, you could use that to modify .py files.

**Reply by Dandy-Don** [reply]
No.  Network is not available.....

**Comment by Dandy-Don** [comment]
I got it working!  I used a Pico Probe to load the .elf file onto the Pico.  

Now I want to try this using the Adafruit ItsyBitsy RP2040.....

Thanks for the help!!
